kernel/sched: Adjust locking in z_swap()

andyross opened the pull request:
[I wrote this as part of work to address a known abort race in #32344 .  But it turns out that this change all by itself hides that problem for me.  It is NOT a fix for the race, but the window (which got opened in the deadlock fix a few commits back) is now so tiny that I can't make it happen.   So it's probably worth reviewing and committing early to unwreck CI.] 

Swap was originally written to use the scheduler lock just to select a
new thread, but it would be nice to be able to rely on scheduler
atomicity later in the process.  Rework the code a bit so that swap
takes the lock itself and holds it until just before the call to
arch_switch().

Note that the local interrupt mask has always been required to be held
across the swap, so extending the lock here has no effect on latency
at all on uniprocessor setups, and even on SMP only affects average
latency and not worst case.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix race between thread wakeup timeout and abort

Aborted threads will cancel their timeouts, but the timeout subsystem
isn't protected under the same lock so it's possible for a timeout to
fire just as a thread is being aborted and wake it up unexpectedly.
Check the state before blowing anything up.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Adjust locking in z_swap()

Swap was originally written to use the scheduler lock just to select a
new thread, but it would be nice to be able to rely on scheduler
atomicity later in the process.  Rework the code a bit so that swap
takes the lock itself and holds it until just before the call to
arch_switch().

Note that the local interrupt mask has always been required to be held
across the swap, so extending the lock here has no effect on latency
at all on uniprocessor setups, and even on SMP only affects average
latency and not worst case.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

jharris-intel commented on kernel/include/kswap.h:
```diff
@@ -62,6 +64,7 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 					  int is_spinlock)
 {
 	ARG_UNUSED(lock);
+	k_spinlock_key_t schedkey = {};
```
Why initialized here? /nit

In my experience initializing variables when they are expected to be overwritten later is an antipattern and tends to lead to even stranger bugs down the line.

(I'm used to late-declaration, which sidesteps this problem in many cases, not to mention often producing better code... but Zephyr's coding style doesn't do that.)

andyross replied:
The compiler isn't always very good at detecting that all paths lead to assignment.  Honestly I don't remember if this was true here or if the initializer just left my fingers unbidden.

jharris-intel replied:
I know what you mean - I hit cases of

```
foo_t *f;
if (complex_expression_a) {
    f = bar();
}
...
if (complex_expression_b) {
    baz(f);
}
```
where it's nontrivial to determine if complex_expression_b actually implies complex_expression_a, all the time. (e.g. if the expressions are both `qux->some_flag`, and the compiler didn't manage to determine that nothing aliased with `some_flag` or `qux`).

...though I'd be mildly surprised if the compiler couldn't detect it in this particular case. There's an unconditional assignment down at line 78.

jharris-intel commented on kernel/sched.c:
```diff
@@ -222,10 +222,8 @@ static ALWAYS_INLINE void dequeue_thread(void *pq,
-			_priq_run_add(&_kernel.ready_q.runq, curr);
-		}
+	if (z_is_thread_queued(curr)) {
+		_priq_run_add(&_kernel.ready_q.runq, curr);
```
Is this change because this spinlock here is now unnecessary, or because we now expect `z_requeue_current` to be called with `sched_spinlock` held?

One thing I will say as an "outsider" here is that it's decidedly non-obvious what functions are intended to be called with/without certain spinlocks held. (Ties into #32148 I suppose.)

(One possible future improvement here would be to add defines for REQUIRES_SPINLOCK_HELD / NOT_HELD that normally are nop-d out but expands to appropriate asserts based on a configuration flag.)

andyross replied:
The latter; lock management is now outside the function in the swap implementation.  And yes, it's a quick change so I didn't bother to do any refactoring to clarify the usage.

Commit:
kernel/thread: Initialize pended_on field of struct thread_base

This got missed, leaving garbage there for restarted threads to trip
on.  Actually I see multiple uninitialized fields, which seems odd.
This code deserves some rework, thread initialization isn't a
performance path and we should probably be zeroing the struct out.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Update to fix the glitch with uniprocessor switch platforms.  Roll in two other obvious and easily-reviewed fixes.

jharris-intel commented on kernel/include/kswap.h:
```diff
@@ -68,11 +70,18 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 	}
+	if (!is_spinlock || lock != &sched_spinlock) {
+		(void) k_spin_lock(&sched_spinlock);
+	}
```
The above logic will (seemingly) misbehave if

```
is_spinlock=true;
lock=NULL;
```

It'll take the `sched_spinlock` but discard interrupt state, likely leaving us interrupt-disabled forever.

If this is expected to never come up, is there a reason for having a separate `is_spinlock` as opposed to just using `NULL` in this case?

If this is expected to come up, what are the desired semantics?

andyross replied:
This is subtle (I mean, it's the guts of the context switch, that's sort of a given).  But in fact no, the **outer** key state that we were passed as an argument is what gets restored.  That is, we're assuming details about the spinlock key implementation: it's the same value as returned by the underlying arch_irq_lock() primitive, which is recursive in this way.  Back in the uniprocessor days, irq_lock() was all we had, so that's the primitive architectures are expected to provide and SMP is built on top of that (and then had to provide a compatible implementation of irq_lock() that worked across CPUs, heh).

This would never be allowed in application code, but I'm OK with it here.

And the bit about "is_spinlock" is a holdover from the same era.  We used to pass an irq_lock() key to (what at the time was called) _Swap().  And the SMP version still supports that, so we actually have two versions of swap, one for spinlocks and one for traditional irq_locks.  So this function gets expanded twice, with is_spinlock used as a compile-time value to select which specific locking strategy we use.

(FWIW: I just checked, there's only two spots in the posix subsystem still using the old primitive.  So we should just port them to spinlocks and remove this.)

jharris-intel commented:
Build failures?

jharris-intel commented:
Hm. I'm hitting an occasional hang with this patch (+ MrVan's SMP patch). Don't know if it was actually caused by this patch or if this patch just changes timing or somesuch.

Hang is with one core spinning in `wait_for_switch` (`z_get_next_switch_handle`->`LOCKED(&sched_spinlock) {` -> `if (old_thread != new_thread) {` -> `wait_for_switch(new_thread);`), and all other cores spinning in various places trying to grab the sched_lock.

What is supposed to happen in this case?

(General comment: feel free to split this - and in general all of my comments - into new issue(s) as appropriate.)

andyross commented:
Closing this one.  All these patches are now folded into the larger series at #32557

andyross commented:
@jharris-intel Indeed, the timing changes did expose other issues.  I believe I'm chasing the last x86_64 one right now, but am not testing on ARM.  See the larger PR for details, but it's not unlikely you have something else going on.

As far as that particular spinning loop: what is supposed to happen is that the spin lasts only for as long as a thread can add itself to the run queue (in requeue_current()) and get to the end of the context switch, there shouldn't be any locking points on that code path in the z_swap() case at all (and it's not an issue for interrupt return or thread initialization, where the switch handle is already present).  Indeed there **was** a deadlock fixed in that dependency a while back, which is why those two steps are in the order they are.

jharris-intel commented:
Thank you the explanation, but I'm still missing something. Perhaps an elaboration may help show what I'm missing here...

Core, say, 0, starts [`z_get_next_switch_handle`](https://github.com/andyross/zephyr/blob/16ab89ea0a0c0bef319d7bbc4f7abe0e8ef023bc/kernel/sched.c#L985). It locks the `sched_spinlock` [here](https://github.com/andyross/zephyr/blob/16ab89ea0a0c0bef319d7bbc4f7abe0e8ef023bc/kernel/sched.c#L992). This spinlock lock also disables interrupts for core 0. Core 0 finds that it needs to switch, [here](https://github.com/andyross/zephyr/blob/16ab89ea0a0c0bef319d7bbc4f7abe0e8ef023bc/kernel/sched.c#L1000), and then enters the loop [here](https://github.com/andyross/zephyr/blob/16ab89ea0a0c0bef319d7bbc4f7abe0e8ef023bc/kernel/sched.c#L1002).

This loop is as follows

```c
static inline void wait_for_switch(struct k_thread *thread)
{
#ifdef CONFIG_SMP
	volatile void **shp = (void *)&thread->switch_handle;

	while (*shp == NULL) {
		k_busy_wait(1);
	}
#endif
}
```

We're sitting in it with interrupts disabled on this core. Which means that if the `while` loop is entered at all, we will spin "forever", or rather until another core comes to our rescue.

However.

All the other cores have tried to grab `sched_spinlock`. The code for grabbing a spinlock first disables interrupts, then spins until the spinlock is available.

So none of the other cores _can_ come to our rescue here. All cores have interrupts disabled and are waiting for an event of some sort (the other cores waiting for `sched_spinlock` to become available, core 0 waiting for `thread->switch_handle` to be written to). And now we're deadlocked.

*****

> the spin lasts only for as long as a thread can add itself to the run queue

I do not follow, my apologies. Where, precisely, can the thread "add itself to the run queue" in this scenario? We have interrupts disabled on all cores at this point, and all cores are spinning waiting for a condition.

Or are you saying that every thread's `switch_handle` must be normally non-null, and only set to NULL inside `sched_spinlock`, and only if it's reset to non-NULL before releasing `sched_spinlock`? Because if so there are several places that appear to not follow that invariant, e.g.

```c
		if (IS_ENABLED(CONFIG_SMP)) {
			/* Active threads MUST have a null here */
			new_thread->switch_handle = NULL;
		}
	}
```
(Note implicit lock release at closing brace.)

*****

My apologies, I'm likely missing something obvious here.
